### PR TITLE
Add mock tool

### PR DESCRIPTION
**Pull request by ljmf00**

### CODE REVIEW

File: docs/descriptions/description.json [hunk 1]
```diff
@@ -0,0 +1,12 @@
+[
```
Comment (andreaTP) [inline docs/descriptions/description.json]: tabs???????? :-1:

File: tool.sh [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+		echo "Can't parse .codacyrc file"
+		exit 1
+	fi
+
+	codacy_files=$(jq -cer '.files | .[]' <<< "$codacyrc_file")
+	codacy_patterns="$(jq -cer ".tools | .[] | select(.name==\"dummy\") | .patterns | .[].patternId" <<< "$codacyrc_file")"
+	codacy_foobar_parameter="$(jq -cer ".tools | .[] | select(.name==\"dummy\") | .patterns | .[].parameters | .[] | select(.name==\"value\") | .value" <<< "$codacyrc_file")"
+
+	if [ $? -ne 0 ]; then
+		unset codacy_foobar_parameter
```
Comment (andreaTP) [inline tool.sh]: can we avoid using globals?
specially resetting them .... it's really hard to reason about

File: tool.sh [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+	fi
+
+	if [ "$codacy_patterns" == "" ]; then
+		# When no supplied patterns
```
Comment (andreaTP) [inline tool.sh]: move the comment before the condition

Comment (andreaTP) [reply tool.sh]: isn't it easier to check the exitcode of `jq` instead of relying on empty string?

File: tool.sh [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+		load_config_file
+	fi
+
+	if [ "$codacy_files" == "" ]; then
+		# When no files given, run with all files in /src
+		load_src_files
+	fi
+else
+	load_config_file
```
Comment (andreaTP) [inline tool.sh]: call these functions just once with an or in the condition removing this `else` branch

File: tool.sh [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+	load_src_files
+fi
+
+function has_pattern {
+  local pattern="$1"
+  if [[ $codacy_patterns =~ (^|[[:space:]])"$pattern"($|[[:space:]]) ]] ; then
+    ret=0
```
Comment (andreaTP) [inline tool.sh]: directly return without passing through a variable to handle the result

File: tool.sh [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+function report_error {
+	local file="$1"
+	local output="$2"
+	echo "{\"filename\":\"$file\",\"message\":\"found $output\",\"patternId\":\"foobar\",\"line\":1}"
+}
+
+while read -r file; do
+	final_file="/src/$file"
+	if [ -f "$final_file" ]; then
```
Comment (andreaTP) [inline tool.sh]: move the body of the `while` to a function

File: README.md [hunk 7]
```diff
@@ -1,2 +1,350 @@
...
+
+* To run the tool in debug mode, so you can have more detailed logs, you need to
+  set the environment variable `DEBUG` to `true` when invoking the docker.
+* To configure a different timeout for the tool, you have to set the environment
+  variable `TIMEOUT` when invoking the docker, setting it with values like
+  `10 seconds`, `30 minutes` or `2 hours`.
+
+#### Setup
+
+1. Write the docker file that will run the tool.
```
[the remark this answers is not in the document]
Comment (ljmf00) [reply README.md]: Done

File: README.md [hunk 8]
```diff
@@ -1,2 +1,349 @@
...
+* UnusedCode
+* Security
+* Compatibility
+* Performance
+* Documentation
+* BestPractice
+
+##### Description structure
+
+The documentation description is optional. So this could be skipped, but try as much as possible adding them.
```
Comment (heliocodacy) [inline README.md]: I would remove the "skipped" part. It's optional but the tool itself will bring way more value if the issue is properly documented

File: README.md [hunk 9]
```diff
@@ -1,2 +1,349 @@
...
+--user=docker \
+--rm=true \
+-v <PATH-TO-FOLDER-WITH-FILES-TO-CHECK>:/src:ro \
+<YOUR-DOCKER-NAME>:<YOUR-DOCKER-VERSION>
+```
+
+**Docker restrictions**:
+
+* Docker image size should not exceed 500MB
+* Docker should contain a non-root user named docker with UID/GID 2004
```
Comment (heliocodacy) [inline README.md]: and group?

File: README.md [hunk 10]
```diff
@@ -1,2 +1,349 @@
-# codacy-example-tool
-Docker engine example for a codacy tool
+# Codacy Example Tool [![Codacy Badge](https://api.codacy.com/project/badge/Grade/7622fdf861c34d69971a8b04aca65fef)](https://www.codacy.com/manual/Codacy/codacy-example-tool?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=codacy/codacy-example-tool&amp;utm_campaign=Badge_Grade) [![Build Status](https://circleci.com/gh/codacy/codacy-example-tool.svg?style=shield&circle-token=:circle-token)](https://circleci.com/gh/codacy/codacy-example-tool)
+
+Docker engine example for a codacy tool.
+
+This tool is an example and test tool for Codacy.
+
```
[the remark this answers is not in the document]
Comment (josemiguelmelo) [reply README.md]: Added the explanation about what is a "Codacy Tool" and referenced it on the rephrased sentence.

File: README.md [hunk 11]
```diff
@@ -1,2 +1,349 @@
...
+## Documentation
+
+### What is a Codacy tool
+
+A codacy tool is a docker container running a wrapper around linters with
+codacy output standards. The tool provides the results through the standard
+output using our format.
```
[the remark this answers is not in the document]
Comment (josemiguelmelo) [reply README.md]: Nice suggestion 👍

File: README.md [hunk 12]
```diff
@@ -1,2 +1,349 @@
...
+> To know more about dockers, and how to write a docker file please refer to
+> [https://docs.docker.com/reference/builder/](https://docs.docker.com/reference/builder/)
+
+You can check the code of an already implemented tool and, if you wish, fork it
+to start your own. You are free to modify and use it for your own tools.
+
+#### Structure
+
+* To run the tool we provide the configuration file, `/.codacyrc`, with the
+  language to run and optional parameters your tool might need.
```
[the remark this answers is not in the document]
Comment (josemiguelmelo) [reply README.md]: Done ✅

File: README.md [hunk 13]
```diff
@@ -1,2 +1,349 @@
...
+  for all folders in /src).
+* If `/.codacyrc` exists and has only files and no patterns, run only for those
+  files and look for the tool's native configuration file, if the tool supports
+  it.
+* If `/.codacyrc` does not exist or any of its contents (files or patterns) is
+  not available, you should invoke the tool for all files from /src (files
+  should be searched recursively for all folders in /src) and check them with
+  the tool's native configuration file, if it is supported and if it exists.
+  Otherwise, run the tool with the default patterns.
+* If `/.codacyrc` fails to be parsed, throw an error.
```
[the remark this answers is not in the document]
Comment (josemiguelmelo) [reply README.md]: Tried to simplify it. What do you think?

File: README.md [hunk 14]
```diff
@@ -1,2 +1,349 @@
...
+
+* ErrorProne
+* CodeStyle
+* Complexity
+* UnusedCode
+* Security
+* Compatibility
+* Performance
+* Documentation
+* BestPractice
```
[the remark this answers is not in the document]
Comment (josemiguelmelo) [reply README.md]: Done ✅

File: README.md (hunk 8 above)
Comment (josemiguelmelo) [reply README.md]: Done ✅

File: README.md [hunk 15]
```diff
@@ -1,2 +1,349 @@
...
+* A single /docs/description/description.json file.
+* A /docs/description/`<PATTERN-ID>.md` file for each pattern.
+
+##### Levels and Categories
+
+For level types we have:
+
+* Error
+* Warning
+* Info
```
[the remark this answers is not in the document]
Comment (josemiguelmelo) [reply README.md]: Done ✅

File: README.md [hunk 16]
```diff
@@ -1,2 +1,349 @@
...
+  should be searched recursively for all folders in /src) and check them with
+  the tool's native configuration file, if it is supported and if it exists.
+  Otherwise, run the tool with the default patterns.
+* If `/.codacyrc` fails to be parsed, throw an error.
+
+##### General tool behavior
+
+**Exit codes**:
+
+* The exit codes can be different, depending if the tool invocation is
```
[the remark this answers is not in the document]
Comment (josemiguelmelo) [reply README.md]: Done ✅

File: .circleci/config.yml [hunk 17]
```diff
@@ -0,0 +1,40 @@
+version: 2.1
+
+orbs:
+  codacy: codacy/base@1.1.0
+  codacy_plugins_test: codacy/plugins-test@0.6.5
```
Comment (lolgab) [inline .circleci/config.yml]: Bump versions here

File: .github/CODEOWNERS [hunk 18]
```diff
@@ -0,0 +1 @@
+* @codacy/toss
```
Comment (lolgab) [inline .github/CODEOWNERS]: This is crazy good! No need to change all repositories when the team changes!

File: Dockerfile [hunk 19]
```diff
@@ -0,0 +1,10 @@
+FROM alpine:3.10.3
```
Comment (lolgab) [inline Dockerfile]: I think minor version is enough here: `alpine:3.10`

File: .github/CODEOWNERS (hunk 18 above)
Comment (josemiguelmelo) [reply .github/CODEOWNERS]: Yeah! We should do it for the rest of the repositories 👌

File: .circleci/config.yml (hunk 17 above)
Comment (josemiguelmelo) [reply .circleci/config.yml]: Done ✅

File: Dockerfile (hunk 19 above)
Comment (josemiguelmelo) [reply Dockerfile]: Yes, you're right. We can use 3.10 instead
Updated it

**Comment by lolgab** [comment]
👏👏👏 Thank you :)
Thank you @ljmf00 for the initial work 👏

**Pull request merged**